K3S Maintenance Doc (#4)
appuio/rancher-docs: docs/modules/ROOT/pages/how-tos/k3s_maintenance.adoc

> = K3S Maintenance

Reviewer: As per https://products.docs.vshn.ch/products/appuio/managed/k3s.html we write `K3s` instead of `K3S`. Please change that in this document.

Author: done

== Kubernetes Maintenance

The Kubernetes version is bound to the K3S version, so by updating K3S you also update the Kubernetes version.
The https://github.com/rancher/system-upgrade-controller[System Upgrade Controller] from Rancher is used to upgrade the K3S version.

> The System Upgrade Controller is installed using a https://github.com/projectsyn/component-system-upgrade-controller[Project Syn component]

Reviewer: ```suggestion
The System Upgrade Controller is installed using a https://github.com/projectsyn/component-system-upgrade-controller[Project Syn component].
```

> In order to manage the K3s version the following System Upgrade Controller plan is used (the plan is also installed using the Project Syn component):

Reviewer 2: ```suggestion
In order to manage the K3s version the following System Upgrade Controller plan is deployed by the Project Syn component:
```

[source,yaml]
----
include::example$suc_plan_k3s.yaml[]
----

> Check the https://github.com/k3s-io/k3s/releases[K3s Release] page and update the value `version` with the version you want to upgrade to. Example: `version: v1.20.4+k3s1`

Reviewer 2: ```suggestion
Check the https://github.com/k3s-io/k3s/releases[K3s Release] page and update the value of `version` with the version you want to upgrade to in the Project Syn hierarchy.
For example, to upgrade to "v1.20.4+k3s1", set `version: v1.20.4+k3s1`.
Don't forget to recompile the cluster catalog to roll out the updated Plan object on the cluster.
```

> This will immediatly trigger an upgrade of K3s. The container image `docker.io/rancher/k3s-upgrade` will replace the existing K3s binary and restart K3s.

Reviewer 2: ```suggestion
Once the plan is updated on the cluster, it will immediately trigger an upgrade of K3s.
The container image `docker.io/rancher/k3s-upgrade` will replace the existing K3s binary and restart K3s.
```

> == Node Maintenance

Reviewer: Not sure if there is another documentation around which describes the SUC process, please ask around internally at VSHN to see if there is something.

Reviewer 2: Not yet, to the best of my knowledge.

> We use Ubuntu as the underlaying operating system for K3s Kubernetes clusters. Node maintenance is also done using the System Upgrade Controller. 

Reviewer 2: ```suggestion
We use Ubuntu as the underlying operating system for K3s Kubernetes clusters.
Node maintenance is also done using the System Upgrade Controller. 
```

> To upgrade the system packages the following scripts in a container image are used: https://github.com/projectsyn/system-upgrade-controller-package-upgrade

Reviewer 2: Probably worth mentioning that we currently build the SUC package upgrade images for Ubuntu Bionic (18.04) and Focal (20.04).

Author: :+1:

> Every week a new container image is build with the updates os packages. This container image is used in the System Upgrade Controller plan.

Reviewer 2: ```suggestion
Every week on Monday morning a new container image is built with the updated OS packages.
This container image is used in the System Upgrade Controller plan.
```

> The following System Upgrade Controller plan is used (the plan is also installed using the Project Syn component):

Reviewer 2: Make a note that the Plan needs to use the correct SUC package upgrade image for the deployed Ubuntu version.

Author: :+1:

[source,yaml]
----
include::example$suc_plan_os.yaml[]
----

> This plan uses a `channel` instead of the `version` argument. The Project Syn component configures this to the https://github.com/projectsyn/floodgate[Floodgate URL]: `https://floodgate.syn.vshn.net/window/<day>/<hour>`

Reviewer 2: ```suggestion
This plan uses a `channel` instead of the `version` argument.
The Project Syn component configures the channel to use https://github.com/projectsyn/floodgate[Floodgate] to schedule a weekly node maintenance window: `https://floodgate.syn.vshn.net/window/<day>/<hour>`
```

> Floodate returns the version of the upgrade container to be used. The time of upgrade is controller with the `day` and `hour` variable. An update to the latest OS package version is only done after this time.

Reviewer 2: ```suggestion
Floodgate returns the version of the upgrade container to be used.
The upgrade window is controlled with the `day` and `hour` variables.
An update to the latest OS package versions is done during the upgrade window.
Note that there's no way to specify the end of the upgrade window, only the start is controlled by the `day` and `hour` variables.
See the https://github.com/projectsyn/component-system-upgrade-controller/blob/master/docs/modules/ROOT/pages/references/parameters.adoc#plans[System Upgrade Controller component documentation] for a more detailed description of the `day` and `hour` variables.
```

> In order to upgrade K3s nodes using the System Upgrade Controller, they have to be labeled with `plan.upgrade.cattle.io/focal`

Reviewer 2: Probably worth mentioning that the label on the nodes must match the label in the `Plan` object.
